backend/src/apiserver/storage/pipeline_store.go

```diff
@@ -425,3 +447,350 @@ func (s *PipelineStore) UpdatePipelineAndVersionsStatus(id string, status model.
+	if err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to create a pipeline version id.")
+	}
+	newPipelineVersion.UUID = id.String()
+
+	// Prepare queries of inserting new version and updating default version.
+	versionSql, versionArgs, versionErr := sq.
+		Insert("pipeline_versions").
+		SetMap(
+			sq.Eq{
+				"UUID":           newPipelineVersion.UUID,
+				"CreatedAtInSec": newPipelineVersion.CreatedAtInSec,
+				"Name":           newPipelineVersion.Name,
+				"Parameters":     newPipelineVersion.Parameters,
+				"PipelineId":     newPipelineVersion.PipelineId,
+				"Status":         string(newPipelineVersion.Status),
+				"CodeSourceUrl":  ""}).
+		ToSql()
+	if versionErr != nil {
+		return nil, util.NewInternalServerError(
+			versionErr,
+			"Failed to create query to insert version to pipeline version table: %v",
+			versionErr.Error())
+	}
+	pipelineSql, pipelineArgs, pipelineErr := sq.
+		Update("pipelines").
+		SetMap(sq.Eq{"DefaultVersionId": newPipelineVersion.UUID}).
+		Where(sq.Eq{"UUID": newPipelineVersion.PipelineId}).
+		ToSql()
+	if pipelineErr != nil {
+		return nil, util.NewInternalServerError(
+			pipelineErr,
+			"Failed to create query to update pipeline default version id: %v",
+			pipelineErr.Error())
+	}
+
+	// In a single transaction, insert new version and update default version.
+	tx, err := s.db.Begin()
+	if err != nil {
+		return nil, util.NewInternalServerError(
+			err,
+			"Failed to start a transaction: %v",
+			err.Error())
+	}
+
+	_, err = tx.Exec(versionSql, versionArgs...)
+	if err != nil {
+		tx.Rollback()
+		if s.db.IsDuplicateError(err) {
+			return nil, util.NewInvalidInputError(
+				"Failed to create a new pipeline version. The name %v already exist. Please specify a new name.", v.Name)
+		}
+		return nil, util.NewInternalServerError(err, "Failed to add version to pipeline version table: %v",
+			err.Error())
+	}
+	_, err = tx.Exec(pipelineSql, pipelineArgs...)
+	if err != nil {
+		tx.Rollback()
+		return nil, util.NewInternalServerError(err, "Failed to update pipeline default version id: %v",
+			err.Error())
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to create new pipeline version: %v",
+			err.Error())
+	}
+
+	return &newPipelineVersion, nil
+}
+
+func (s *PipelineStore) UpdatePipelineDefaultVersion(pipelineId string, versionId string) error {
+	sql, args, err := sq.
+		Update("pipelines").
+		SetMap(sq.Eq{"DefaultVersionId": versionId}).
+		Where(sq.Eq{"UUID": pipelineId}).
+		ToSql()
+	if err != nil {
+		return util.NewInternalServerError(err, "Failed to create query to update the pipeline default version: %s", err.Error())
+	}
+	_, err = s.db.Exec(sql, args...)
+	if err != nil {
+		return util.NewInternalServerError(err, "Failed to update the pipeline default version: %s", err.Error())
+	}
+
+	return nil
+}
+
+func (s *PipelineStore) GetPipelineVersion(versionId string) (*model.PipelineVersion, error) {
```

Comment:
can you remind me why we need two methods? get and get status

Reply:
the latter is used in test to quickly fetch a pipeline (or pipeline version) in certain status? e.g., when testing deletion, db is updated, minio failed; and pipeline in db should be in deleting status. so get pipeline or pipeline version with deleting status to verify.